enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+
+#### Pre-release / Development Feature Gate
+
+In order to achieve desired cluster configuration for cloud components during development and technical preview, we will add a new OpenShift [FeatureGate](https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html) resource named `ExternalCloudProvider`, which will allow developers and customers to force their cluster to upgrade from in-tree to out-of-tree cloud providers.
+
+This feature gate will be shipped with `TechPreviewNoUpgrade` `FeatureSet`, preventing upgrades.
+
+The feature gate will be used to determine whether or not operators should migrate their components to the out-of-tree configuration.
+If the feature gate is found, the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for current platform.
+
+Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the feature gate, allowing production clusters to automatically upgrade to the out-of-tree implementation during 4.N to 4.N+1 upgrade.
+
+Once all providers are moved out-of-tree, the feature gate will be safe to remove.
+
+This feature gate will be used in CI tasks, to test migration stability.
+
+#### Migration procedure
+
+The process of the migration will be managed by corresponding operators. Communication will be done via `CloudControllerOwner` condition on the operator's `ClusterOperator` resource.
+
+```yaml
+conditions:
+- type: CloudControllerOwner
+  status: True
+```
+
+`status` will indicate the state of the ownership for this code. Setting it to `False` on one `ClusterOperator` will unblock resource provisioning for the other operator.
+
+The plan for migration is the following:
+
+1) `kubelet` and `kube-controller-manager`.
+    1) Set the `CloudControllerOwner` condition on the `ClusterOperator` with `status=True`.
+    2) Disable cloud loops in its operands.
+    3) Wait for the new resource version rollout to complete and become stable.
+    4) Indicate the completion of the migration by setting `status=False` on the `CloudControllerOwner` condition.
+2) `cloud-controller-manager`
+    1) Set the `CloudControllerOwner` condition on the `ClusterOperator` with `status=False`.
+    2) Wait for preceding operators to have the `CloudControllerOwner` condition with `status=False`.
+    3) Provision out-of-tree resources.
+    4) Indicate the ownership transfer by setting `status=True` on the `CloudControllerOwner` condition.
+    3) Wait for the resources rollout to finish.
+
+Once all resources are up and running, the migration is completed.
```

Comment:
There should be a short note somewhere about default values - when running an older kubelet or controller-manager that does not know anything about external cloud providers yet, missing `CloudControllerOwner` condition should tell the world around (CCCMO?) that in-tree cloud provider is used there.